workers/unity/Packages/com.improbable.gdk.core/Systems/SpatialOSReceiveSystem.cs

```diff
@@ -54,33 +55,68 @@ protected override void OnDestroyManager()
 
         protected override void OnUpdate()
         {
-            var updateSystem = World.GetExistingManager<ComponentUpdateSystem>();
             if (worker.Connection == null)
             {
                 return;
             }
 
-            do
+            try
             {
-                using (var opList = worker.Connection.GetOpList(0))
+                do
                 {
-                    try
-                    {
-                        Dispatcher.Process(opList);
-                    }
-                    catch (Exception e)
+                    using (var opList = worker.Connection.GetOpList(0))
                     {
-                        worker.LogDispatcher.HandleLog(LogType.Exception, new LogEvent("Exception:")
-                            .WithException(e));
+                        inCriticalSection = opDeserializer.ParseOpListIntoDiff(opList, diff);
                     }
                 }
+                while (inCriticalSection);
+
+                inCriticalSection = false;
+                opDeserializer.Reset();
+
+                if (diff.Disconnected)
+                {
+                    OnDisconnect(diff.DisconnectMessage);
+                    return;
+                }
+
+                foreach (var entityId in diff.GetEntitiesAdded())
+                {
+                    AddEntity(entityId);
+                }
+
+                World.GetExistingManager<ComponentUpdateSystem>().ApplyDiff(diff);
```

Comment:
These managers should be cached elsewhere before using here

Reply:
They are going to be removed from here shortly so I hadn't given this much thought. But sure they should be cached in the meantime.